migrate-components.sh

```diff
@@ -0,0 +1,125 @@
+#!/bin/bash
+
+# Migrates the spec.containerImage field to status.lastPromotedImage if
+# status.lastPromotedImage is not already set.
+#
+# If spec.containerImage has been updated by imageRepository (has not sha)
+# then the script will attempt to get the latest push snapshot or override
+# snapshot that contains the component and gets the image from there
+#
+# REQUIREMENTS:
+#  - kubectl >= 1.24
+#  - jq
+
+younger() {
+    RESOURCE_A=$1
+    RESOURCE_B=$2
+
+    # compare metadata.creationTimestamp and return whichever is younger
+}
+
+# takews two snapshot jsons and returns whichever has the newest metadata.creationTimestamp
+get_newest_snapshot_json() {
+    PUSH_SNAPSHOT_JSON=$1
+    OVERRIDE_SNAPSHOT_JSON=$2
+
+    # otherwise, return whichever is younger
+    PS_TIMESTAMP=$(echo $PUSH_SNAPSHOT_JSON | jq .metadata.creationTimestamp)
+    OS_TIMESTAMP=$(echo $OVERRIDE_SNAPSHOT_JSON | jq .metadata.creationTimestamp)
+    if [[ $OS_TIMESTAMP > $PS_TIMESTAMP ]]; then
+        echo $OVERRIDE_SNAPSHOT
+    fi
+    echo $PUSH_SNAPSHOT
+}
+
+get_latest_override_snapshot_with_component() {
+    COMPONENT=$1
+    NAMESPACE=$2
+    APPLICATION=$3
+
+    SNAPSHOTS=$(kubectl get snapshots -n $NAMESPACE --sort-by .metadata.creationTimestamp --selector=pac.test.appstudio.openshift.io/type=override,appstudio.openshift.io/application=$APPLICATION --no-headers | tac | cut -d ' ' -f1)
+
+    if ! [[ $SNAPSHOTS ]]; then
+        return 0
+    fi
+
+    while read -r snapshot; do
+        snapshot_json=$(kubectl get -o json snapshot $snapshot -n $NAMESPACE)
+        echo "all snapshots: $SNAPSHOTS"
```

Comment:
Do we still wanna list all snapshots? Some namespaces have over 1000 of them.

Reply:
That and the line above were debugging output, I removed them.  Just testing the change locally now